lib/routes/account.js

```diff
@@ -1155,68 +1155,77 @@ module.exports = (log, db, mailer, Password, config, customs, signinUtils, push)
-          .then(
-            function (emailRecord) {
-              uid = emailRecord.uid
-
-              const password = new Password(authPW, emailRecord.authSalt, emailRecord.verifierVersion)
-              return signinUtils.checkPassword(emailRecord, password, request.app.clientAddress)
-                .then(
-                  function (match) {
-                    if (! match) {
-                      throw error.incorrectPassword(emailRecord.email, form.email)
-                    }
-                    // We fetch the devices to notify before deleteAccount()
-                    // because obviously we can't retrieve the devices list after!
-                    return db.devices(uid)
-                  }
-                )
-                .then(
-                  function (devices) {
-                    devicesToNotify = devices
-                    return db.deleteAccount(emailRecord)
-                  }
-                )
-                .then(
-                  function () {
-                    push.notifyAccountDestroyed(uid, devicesToNotify)
-                      .catch(() => {})
-                    return P.all([
-                      log.notifyAttachedServices('delete', request, {
-                        uid: uid,
-                        iss: config.domain
-                      }),
-                      request.emitMetricsEvent('account.deleted', {
-                        uid: uid
-                      })
-                    ])
-                  }
-                )
-                .then(
-                  function () {
-                    return {}
-                  }
-                )
-            },
-            function (err) {
-              if (err.errno === error.ERRNO.ACCOUNT_UNKNOWN) {
-                customs.flag(request.app.clientAddress, {
-                  email: form.email,
-                  errno: err.errno
-                })
-              }
-              throw err
+          .then((emailRecord) => {
+            return totpUtils.hasTotpToken(emailRecord)
+              .then((hasToken) => {
+                const sessionToken = request.auth && request.auth.credentials
+
+                // Someone tried to delete an account with TOTP but did not specify a session.
+                // This shouldn't happen in practice, but just in case we throw unverified session.
+                if (! sessionToken && hasToken) {
+                  throw error.unverifiedSession()
+                }
+
+                // Ensure that this session has been verified with the correct assurance level for account
+                // if it has TOTP enabled
+                if (sessionToken && hasToken && (sessionToken.tokenVerificationId || sessionToken.authenticatorAssuranceLevel <= 1)) {
+                  throw error.unverifiedSession()
+                }
+
+                // If TOTP is not enabled, we fall back to the default behavior and let the user
```

Comment:
The `If TOTP is not enabled` part of this comment seems misleading, don't we also fall back if TOTP is enabled and the session is TOTP-verified?